3.8/graphs-custom-pregel.md

```diff
@@ -0,0 +1,1535 @@
+
+```json
+"outDegree": {
+  "accumulatorType": "store",
+  "valueType": "ints"
+}
+```
+
+A vertex knows exactly how many outgoing edges it has by definition. Therefore
+we only have to set the amount to an accumulator once and not multiple times.
+With that knowledge it makes sense to set the `accumulatorType` to store, as
+no further calculations need to take place. As the possible amount of
+outgoing edges is even, we are setting `valueType` to `ints`. Additionally, we
+do not need to store the sender as that value is out of interest in that example.
+
+##### inDegree
+
+What a vertex does **not know** is, how many incoming edges it has.
+Therefore we need to get them to know that value.
+
+```json
+"inDegree": {
+  "accumulatorType": "sum",
+  "valueType": "ints"
+}
+```
+
+The choice for `valueType` is equal compared to "outDegree"
+because of the same reason. But as you can see, the `accumulatorType` is now
+set to `sum`. As our vertices do not know how many incoming edges there are,
+each vertex needs to send a message to all outgoing. In our program, a message
+will be sent to every neighbor. That means a vertex with **n** neighbors,
+will send **n** messages. As in our case, a single message represents a single
+incoming edge, we need to add **"+1"** to our accumulator per each message, and
+therefor set the `accumulatorType` to `sum`.
+
+#### Program
+
+##### initProgram
+
+```js
+initProgram: [
+  "seq",
+
+  ["accum-set!", "outDegree", ["this-outbound-edges-count"]], // Sets our outDegree accumulator with ["this-outbound-edges-count"]
+
+  // Init in degree to 0
+  ["accum-set!", "inDegree", 0],             // Initializes our inDegree (sum) accumulator to 0
+  ["send-to-all-neighbours", "inDegree", 1]  // Sends value: "1" to all neighbors, so their inDegree can be raised next round!
+]
+```
+
+##### updateProgram
+
+```
+updateProgram: ["seq",
+  false]
+}]
+```
+
+As in our case, we do not need an update program. We are just inserting a dummy
+(void) method (will be improved in further state). But currently it is necessary,
+because our update program needs to run once to accumulate the inDegrees values
+that have been sent out in our `initProgram`. Therefore `maxGSS` is set to `2`.
+
+#### Storing the result
+
+To be able to store the result, we either need to define a `resultField`
+(which will just store all accumulators into the given resultField attribute)
+or create a `<program>` which will take care of our store procedure.
+The next code snippet demonstrates how a store program could look like:
+
+```js
+1. "dataAccess": {
+2.   "writeVertex": [
+3.     "attrib-set",
+4.       ["attrib-set", ["dict"], "inDegree", ["accum-ref", "inDegree"]], // 1st parameter of outer attrib-set
+5.       "outDegree",                                                     // 2nd parameter of outer attrib-set
+6.       ["accum-ref", "outDegree"]                                       // 3rd parameter of outer attrib-set
+7.   ]
+8. }
+```
```

Comment:
Again, the lines are too long. Also, where are the line numbers coming from?
`````suggestion
```js
"dataAccess": {
  "writeVertex": [
    "attrib-set",
      // 1st parameter of outer attrib-set
      ["attrib-set", ["dict"], "inDegree", ["accum-ref", "inDegree"]],
      // 2nd parameter of outer attrib-set
      "outDegree",
      // 3rd parameter of outer attrib-set
      ["accum-ref", "outDegree"]
  ]
}
```
`````